tracing/src/main/java/com/palantir/tracing/Tracer.java

```diff
@@ -374,6 +375,43 @@ static String getRequestId(DetachedSpan detachedSpan) {
         throw new SafeIllegalStateException("Unknown span type", SafeArg.of("detachedSpan", detachedSpan));
     }
 
+    @Nullable
+    static <T> T getTraceLocalValue(TraceLocal<T> traceLocal) {
+        Trace maybeCurrentTrace = currentTrace.get();
+        if (maybeCurrentTrace == null) {
+            return null;
+        }
+
+        return (T) maybeCurrentTrace.getTraceState().getTraceLocals().get(traceLocal);
+    }
+
+    @Nullable
+    static <T> T getTraceLocalValue(TraceLocal<T> traceLocal, Supplier<T> initialValue) {
+        Trace maybeCurrentTrace = currentTrace.get();
+        if (maybeCurrentTrace == null) {
+            return null;
+        }
+
+        return (T) maybeCurrentTrace
+                .getTraceState()
+                .getTraceLocals()
+                .computeIfAbsent(traceLocal, _key -> initialValue.get());
+    }
+
+    static <T> void setTraceLocalValue(TraceLocal<T> traceLocal, @Nullable T value) {
+        Trace maybeCurrentTrace = currentTrace.get();
+        if (maybeCurrentTrace == null) {
+            return;
+        }
+
+        Map<TraceLocal<?>, Object> traceLocals = maybeCurrentTrace.getTraceState().getTraceLocals();
+        if (value == null) {
+            traceLocals.remove(traceLocal);
+        } else {
+            traceLocals.put(traceLocal, value);
+        }
+    }
```

Comment:
I think we can align the code with intent a bit better by exposing a package-private nullable getter for`TraceState` here, then implemeting the rest of this logic within TraceLocal.

Reply:
it increases the potential for concurrency problems though I think - i.e. without being able to use computeIfAbsent on the ConcurrentHashMap

Reply:
could make tracelocal get synchronized?

Reply:
My preference would be to avoid splitting the complexity between classes, so `Tracer` provides an internal package-private `TraceState getTraceState()` and `TraceState` provides `@Nullable Map<> getTraceLocals()` and `Map<> getOrCreateTraceLocals()`. Then TraceLocal itself can do any threadsafe logic it likes on the map.